### PR TITLE
[SPARK-15334][SQL] HiveClient facade not compatible with Hive 0.12

### PR DESCRIPTION
**Pull request by clockfly**
## What changes were proposed in this pull request?

HiveClient facade is not compatible with Hive 0.12.

This PR Fixes the following compatibility issues:
1. `org.apache.spark.sql.hive.client.HiveClientImpl` use `AddPartitionDesc(db, table, ignoreIfExists)` to create partitions, however, Hive 0.12 doesn't have this constructor for `AddPartitionDesc`.
2. `HiveClientImpl` uses `PartitionDropOptions` when dropping partition, however, class `PartitionDropOptions` doesn't exist in Hive 0.12.
3. Hive 0.12 doesn't support adding permanent functions. It is not valid to call `org.apache.hadoop.hive.ql.metadata.Hive.createFunction`, `org.apache.hadoop.hive.ql.metadata.Hive.alterFunction`, and `org.apache.hadoop.hive.ql.metadata.Hive.alterFunction`
4. `org.apache.spark.sql.hive.client.VersionsSuite` doesn't have enough test coverage for different hive versions 0.12, 0.13, 0.14, 1.0.0, 1.1.0, 1.2.0.
## How was this patch tested?

Unit test.

### CODE REVIEW

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/client/VersionsSuite.scala [hunk 1]
```diff
@@ -130,109 +124,343 @@ class VersionsSuite extends SparkFunSuite with Logging {
...
+      CatalogTable(
+        identifier = TableIdentifier(tableName, Some(database)),
+        tableType = CatalogTableType.MANAGED,
+        schema = Seq(CatalogColumn("key", "int")),
+        storage = CatalogStorageFormat(
+          locationUri = None,
+          inputFormat = Some(classOf[org.apache.hadoop.mapred.TextInputFormat].getName),
+          outputFormat = Some(
+            classOf[org.apache.hadoop.hive.ql.io.HiveIgnoreKeyTextOutputFormat[_, _]].getName),
+          serde = Some(classOf[org.apache.hadoop.hive.serde2.`lazy`.LazySimpleSerDe].getName()),
```
Comment (liancheng) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/client/VersionsSuite.scala]: How about import these Hive classes above instead of using the fully qualified names here?

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/client/VersionsSuite.scala [hunk 2]
```diff
@@ -130,109 +124,343 @@ class VersionsSuite extends SparkFunSuite with Logging {
...
+    test(s"$version: alterTable(tableName: String, table: CatalogTable)") {
+      val newTable = client.getTable("default", "src").copy(properties = Map("changedAgain" -> ""))
+      client.alterTable("src", newTable)
+      assert(client.getTable("default", "src").properties.contains("changedAgain"))
     }
 
-    test(s"$version: set command") {
-      client.runSqlHive("SET spark.sql.test.key=1")
+    test(s"$version: listTables(database)") {
+      assert(client.listTables("default") === Seq("src", "temporary"))
```
Comment (liancheng) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/client/VersionsSuite.scala]: This test case can be flaky because it depends on execution order of other test cases.

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/client/VersionsSuite.scala [hunk 3]
```diff
@@ -130,109 +124,343 @@ class VersionsSuite extends SparkFunSuite with Logging {
...
+    }
+
+    test(s"$version: listTables(database, pattern)") {
+      assert(client.listTables("default", pattern = "src") === Seq("src"))
+      assert(client.listTables("default", pattern = "nonexist") === Seq.empty[String])
```
Comment (liancheng) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/client/VersionsSuite.scala]: Nit: `.isEmpty` is enough for this assert.

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/client/VersionsSuite.scala [hunk 4]
```diff
@@ -130,109 +124,343 @@ class VersionsSuite extends SparkFunSuite with Logging {
...
+
+    val storageFormat = CatalogStorageFormat(
+      locationUri = None,
+      inputFormat = None,
+      outputFormat = None,
+      serde = None,
+      compressed = false,
+      serdeProperties = Map.empty)
+
+
```
Comment (liancheng) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/client/VersionsSuite.scala]: Remove extra newline.

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/client/VersionsSuite.scala [hunk 5]
```diff
@@ -130,109 +124,343 @@ class VersionsSuite extends SparkFunSuite with Logging {
...
+    test(s"$version: getPartitions(catalogTable)") {
+      assert(2 == client.getPartitions(client.getTable("default", "src_part")).size)
     }
 
     test(s"$version: getPartitionsByFilter") {
-      client.getPartitionsByFilter(client.getTable("default", "src_part"), Seq(EqualTo(
-        AttributeReference("key", IntegerType, false)(NamedExpression.newExprId),
+      // Only one partition [1, 1] for key2 == 1
+      val result = client.getPartitionsByFilter(client.getTable("default", "src_part"), Seq(EqualTo(
+        AttributeReference("key2", IntegerType, false)(NamedExpression.newExprId),
```
Comment (liancheng) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/client/VersionsSuite.scala]: Nit: You can omit `NamedExpression.newExprId` and just use an empty parenthesis here.

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/client/HiveShim.scala [hunk 6]
```diff
@@ -208,7 +286,7 @@ private[client] class Shim_v0_12 extends Shim with Logging {
       predicates: Seq[Expression]): Seq[Partition] = {
     // getPartitionsByFilter() doesn't support binary comparison ops in Hive 0.12.
     // See HIVE-4888.
-    logDebug("Hive 0.12 doesn't support predicate pushdown to metastore. " +
+    logWarning("Hive 0.12 doesn't support predicate pushdown to metastore. " +
```
Comment (yhuai) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/client/HiveShim.scala]: For this method , it is allowed to have false positives. Let's update the doc of this method. Also, let's revert this change (just to avoid of letting users think that this warning indicates an error).

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/client/HiveShim.scala [hunk 7]
```diff
@@ -265,6 +343,47 @@ private[client] class Shim_v0_12 extends Shim with Logging {
     dropIndexMethod.invoke(hive, dbName, tableName, indexName, true: JBoolean)
   }
 
+  // Hive 0.12 doesn't support permanent functions. Here we use a in-memory map to simulate the
+  // permanent function registry.
+  private var functionMap = Map.empty[String, CatalogFunction]
+  private val functionDDLNotSupported = "Hive 0.12 doesn't support creating permanent functions. " +
+    "Please use Hive 0.13 or higher."
```
Comment (yhuai) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/client/HiveShim.scala]: For Hive 0.12, let's throw an exception if users want to create a permanent function.

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/client/VersionsSuite.scala [hunk 8]
```diff
@@ -130,109 +124,343 @@ class VersionsSuite extends SparkFunSuite with Logging {
...
-              classOf[org.apache.hadoop.hive.ql.io.HiveIgnoreKeyTextOutputFormat[_, _]].getName),
-            serde = Some(classOf[org.apache.hadoop.hive.serde2.`lazy`.LazySimpleSerDe].getName()),
-            compressed = false,
-            serdeProperties = Map.empty
-          ))
-
-      client.createTable(table, ignoreIfExists = false)
+      client.createTable(table("default", tableName = "src"), ignoreIfExists = false)
+      // Creates a temporary table
+      client.createTable(table("default", "temporary"), ignoreIfExists = false)
```
Comment (yhuai) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/client/VersionsSuite.scala]: I am confused. It is just another table, right? Not a temp table.

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/client/VersionsSuite.scala [hunk 9]
```diff
@@ -130,109 +124,343 @@ class VersionsSuite extends SparkFunSuite with Logging {
...
+    }
+
+    test(s"$version: dropTable") {
+      client.dropTable("default", tableName = "temporary", ignoreIfNotExists = false)
+      assert(client.listTables("default") === Seq("src"))
     }
 
-    test(s"$version: create partitioned table DDL") {
-      client.runSqlHive("CREATE TABLE src_part (value INT) PARTITIONED BY (key INT)")
-      client.runSqlHive("ALTER TABLE src_part ADD PARTITION (key = '1')")
```
Comment (yhuai) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/client/VersionsSuite.scala]: Do we still have this line?

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/client/VersionsSuite.scala (hunk 8 above)
Comment (clockfly) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/client/VersionsSuite.scala]: es, it is another table. We are testing the "drop table", maybe I should rename it to tableToBeDropped.

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/client/HiveShim.scala (hunk 7 above)
Comment (clockfly) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/client/HiveShim.scala]: OK

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/client/VersionsSuite.scala (hunk 9 above)
Comment (clockfly) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/client/VersionsSuite.scala]: SQL "alter table xx add partition" is replaced by hiveClient.createPartitions().

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/client/VersionsSuite.scala [hunk 10]
```diff
@@ -130,109 +124,343 @@ class VersionsSuite extends SparkFunSuite with Logging {
...
+      client.setOut(new PrintStream(new ByteArrayOutputStream()))
+    }
+
+    test(s"$version: setInfo") {
+      client.setInfo(new PrintStream(new ByteArrayOutputStream()))
+    }
+
+    test(s"$version: setError") {
+      client.setError(new PrintStream(new ByteArrayOutputStream()))
+    }
```
Comment (yhuai) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/client/VersionsSuite.scala]: For `setOut`/`setInfo`/`setError`, it is fine to not have tests. They are not actually calling any method of hive metastore client.

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/client/VersionsSuite.scala [hunk 11]
```diff
@@ -130,109 +124,343 @@ class VersionsSuite extends SparkFunSuite with Logging {
...
+
+    test(s"$version: addJar") {
+      client.addJar(".")
```
Comment (yhuai) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/client/VersionsSuite.scala]: Will this actually add anything?

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/client/VersionsSuite.scala [hunk 12]
```diff
@@ -130,109 +124,343 @@ class VersionsSuite extends SparkFunSuite with Logging {
...
+    }
+
+    test(s"$version: newSession") {
+      val newClient = client.newSession()
+      assert(newClient != null)
+    }
+
+    test(s"$version: withHiveState") {
+      client.withHiveState{Thread.currentThread().getContextClassLoader}
```
Comment (yhuai) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/client/VersionsSuite.scala]: For this one, we can check the classloader is `client.clientLoader.classLoader`.

**Comment by yhuai** [comment]
I have finished my review. Thank you for working on this!

**Comment by liancheng** [comment]
LGTM, merging to master and branch-2.0.
